transport/tlscommon/tls.go

```diff
@@ -106,49 +105,49 @@ func ReadPEMFile(log *logp.Logger, s, passphrase string) ([]byte, error) {
 		return nil, err
 	}
 
+	// decrypted, PEM encoded blocks
+	buffer := bytes.NewBuffer(nil)
+
 	for len(content) > 0 {
 		var block *pem.Block
 
 		block, content = pem.Decode(content)
-		if block == nil {
-			if len(blocks) == 0 {
-				return nil, errors.New("no pem file")
-			}
-			break
-		}
 
 		switch {
-		case x509.IsEncryptedPEMBlock(block): //nolint: staticcheck // deprecated, we have to get rid of it
+		case x509.IsEncryptedPEMBlock(block): // nolint: staticcheck // deprecated, we have to get rid of it
 			block, err := decryptPKCS1Key(*block, pass)
 			if err != nil {
 				log.Errorf("Dropping encrypted pem block with private key, block type '%s': %s", block.Type, err)
 				continue
 			}
-			blocks = append(blocks, &block)
+			// it's already a PEM block.
+			buffer.Write(block.Bytes)
 		case block.Type == "ENCRYPTED PRIVATE KEY":
 			block, err := decryptPKCS8Key(*block, pass)
 			if err != nil {
 				log.Errorf("Dropping encrypted pem block with private key, block type '%s', could not decypt as PKCS8: %s", block.Type, err)
 				continue
 			}
-			blocks = append(blocks, &block)
+			// not in PEM format, encode it.
+			err = pem.Encode(buffer, &block)
+			if err != nil {
+				log.Errorf("Dropping encrypted pem block with private key, block type '%s' could not PEM encode it: %s", block.Type, err)
+				continue
+			}
 		default:
-			blocks = append(blocks, block)
+			// not encrypted and not in PEM format, encode it.
```

Comment:
Is this function only used for reading PEM encoded private keys?  The implementation suggests that, but the function name and signature don't.  If this were to read a PEM encoded certificate, you don't want to re-encode.

As it stands, I think the default case is trying to catch if they actually supplied a DER encoded private key.  If that is the case it is probably worth checking to see if `ParsePKIXPublicKey` works and to log at the warn level that DER format was provided but PEM format was expected.